ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,167 @@
+// Libraries
+import {sortBy} from 'lodash'
+
+// Utils
+import {asAssignment, getAllVariables} from 'src/variables/selectors'
+import {buildVarsOption} from 'src/variables/utils/buildVarsOption'
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {runQuery} from 'src/shared/apis/query'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, Variable} from 'src/types'
+
+export const TIME_INVALIDATION = 5000
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (rawText: string): string => {
+  let hash = 0
+  if (!rawText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < rawText.length; i++) {
+    hash = (hash << 5) - hash + rawText.charCodeAt(i)
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  private cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      // TODO(ariel): need to implement specific rules for custom time ranges
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (
+    id: string,
+    hashedVariables: string
+  ): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].hashedVariables !== hashedVariables) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    return this.cache[id].values
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    values: RunQueryResult,
+    isCustomTime: boolean = false
+  ): void => {
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      hashedVariables,
+      isCustomTime,
+      values,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this.cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
```

Comment:
Since we refrain from using expired queries in the `getFromCache` function, I'm leaning towards this being unnecessary?

Reply:
This is used to ensure that the browser doesn't crash due to a lack of memory. The only time we explicitly expire (and delete) from the cache (besides this) is when we submit a query for an existing cached query, or for a query that is matching but has different variables